llvm/lib/CAS/MappedFileRegionBumpPtr.cpp

```diff
@@ -224,11 +224,17 @@ int64_t MappedFileRegionBumpPtr::allocateOffset(uint64_t AllocSize) {
   AllocSize = alignTo(AllocSize, getAlign());
   int64_t OldEnd = BumpPtr->fetch_add(AllocSize);
   int64_t NewEnd = OldEnd + AllocSize;
-  if (LLVM_UNLIKELY(NewEnd > (int64_t)capacity())) {
-    // Try to return the allocation.
-    (void)BumpPtr->compare_exchange_strong(OldEnd, NewEnd);
-    report_fatal_error(
-        errorCodeToError(std::make_error_code(std::errc::not_enough_memory)));
+  if (LLVM_UNLIKELY(NewEnd >= (int64_t)capacity())) {
+    // Return the allocation. If the start already passed the end, that means
+    // some other concurrent allocations already consumed all the capacity.
+    // There is no need to return the original value. If the start was not
+    // passed the end, current allocation certainly bumped it passed the end.
+    // All other allocation afterwards must have failed and current allocation
+    // is in charge of return the allocation back to a valid value.
+    if (OldEnd < (int64_t)capacity())
+      (void)BumpPtr->exchange(OldEnd);
+
+    return 0;
```

Comment:
I try to remove the fatal error but return 0 might actually cause the crash anyway unless we update all the allocation path in the OnDisk format to propagate errors, which I can try but I am not sure how much performance cost is that.